systemd: 253.5 -> 254.3

RaitoBezarius opened the pull request:
###### Description of changes

- brings systemd to v254
- multiple fixes as always
- `pkgsMusl.systemdMinimal` builds, the full one, doesn't

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

RaitoBezarius committed:
systemdMinimal: expose `withRepart` flag and disable it

systemd-repart relies on cryptsetup to perform LUKS operations,
therefore, it's not suitable for systemdMinimal.

RaitoBezarius committed:
systemd: 253.5 -> 254-rc1

This is an early preparation for systemd v254 which causes some patch reflows
and EFI-related cleanups to their new build system with elf2efi, requiring pyelftools
as a Python packge.

RaitoBezarius commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -320,6 +305,9 @@ stdenv.mkDerivation (finalAttrs: {
           { name = "libp11-kit.so.0"; pkg = opt (withHomed || withCryptsetup) p11-kit; }
+
+          # Password quality support
+          { name = "libpasswdqc.so.1"; pkg = opt withPasswordQuality null; }
```
please @arianvp can you package passwdqc ? :P

RaitoBezarius replied:
I will disable it for the time being behind an assert.

flokli commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -141,6 +142,7 @@ assert withCoredump -> withCompression;
 assert withHomed -> withCryptsetup;
 assert withHomed -> withPam;
 assert withUkify -> withEfi;
+assert withRepart -> withCryptsetup;
```
Thanks for adding the assertion :-)

RaitoBezarius committed:
systemd: 254-rc1 -> 254-rc2

nikstur commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -89,6 +88,7 @@
 , withCompression ? true  # adds bzip2, lz4, xz and zstd
 , withCoredump ? true
 , withCryptsetup ? true
+, withRepart ? true
```
If you want to minimize it further, you can add a `withSysupdate` flag ;)

RaitoBezarius replied:
I guess :>

RaitoBezarius commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -527,12 +517,9 @@ stdenv.mkDerivation (finalAttrs: {
 
     "-Defi=${lib.boolToString withEfi}"
-    "-Dgnu-efi=${lib.boolToString withEfi}"
+    "-Dbootloader=${lib.boolToString withEfi}"
```
Hm, yes for sure :)

arianvp commented on pkgs/os-specific/linux/systemd/0002-Don-t-try-to-unmount-nix-or-nix-store.patch:
```diff
@@ -1,18 +1,20 @@
-From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
-From: Eelco Dolstra <eelco.dolstra@logicblox.com>
-Date: Fri, 12 Apr 2013 13:16:57 +0200
+From 400dfb2405f42da99c37a9a5c00edd2dd56e7cc1 Mon Sep 17 00:00:00 2001
```
Hmm could we keep the original commit authors in these patches please?

arianvp replied:
Nevermind. Original-Author is still there

RaitoBezarius replied:
Let me know if this is not enough. It's a bit annoying to go ahead and change all the patches manually to the original author because I was using git to reflow everything.

flokli reviewed (dismissed):
Please reflow the patches with the process described right above the `patches =` line:

https://github.com/NixOS/nixpkgs/pull/243242/files#diff-a284772b505bd573f834b58591d380128545257d9e02601dbe112548526c8710R170-R174

This does also take care of some tiny details like not adding the git version pointers to the patches.

RaitoBezarius committed:
systemd: 254-rc2 -> 254-rc3

https://github.com/systemd/systemd/commit/739bbc58693473b6daa5ab4302f123d2c5fd04ce

RaitoBezarius committed:
systemd: 254-rc3 -> 254

RaitoBezarius committed:
systemd: introduce `withBootloader` for sd-boot, sd-stub, sd-addon

Instead of relying blindly on `withEfi` which may not always be relevant.

RaitoBezarius committed:
systemd: make `systemd-sysupdate` optional via `withSysupdate`

Disable it in `systemdMinimal`

RaitoBezarius committed:
systemd: make `withPasswordQuality` always disabled

It is not possible currently to compile systemd with it because passwdqc is not packaged.

RaitoBezarius commented:
TODO:

- [ ] investigate `hibernate` test
- [ ] run it on actual machine:
   - [ ] laptop
   - [ ] server
- [ ] reflow patches properly
- [ ] more tests?

ElvishJerricco commented:
We will also need to mention in the changelog that `boot.resumeDevice` now must be set to use hibernation when not EFI booting.

ElvishJerricco commented:
These changes are also required:

```diff
diff --git a/nixos/modules/system/boot/systemd/initrd.nix b/nixos/modules/system/boot/systemd/initrd.nix
index 3f40a5b2dfa0..7476a4f9040a 100644
--- a/nixos/modules/system/boot/systemd/initrd.nix
+++ b/nixos/modules/system/boot/systemd/initrd.nix
@@ -57,7 +57,6 @@ let
     "systemd-ask-password-console.service"
     "systemd-fsck@.service"
     "systemd-halt.service"
-    "systemd-hibernate-resume@.service"
     "systemd-journald-audit.socket"
     "systemd-journald-dev-log.socket"
     "systemd-journald.service"
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 5e28c064212e..2787a5d0d970 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -28481,6 +28481,7 @@ with pkgs;
     withTpm2Tss = false;
     withUserDb = false;
     withUkify = false;
+    withBootloader = false;
   };
   systemdStage1 = systemdMinimal.override {
     pname = "systemd-stage-1";
```

RaitoBezarius commented:
> These changes are also required:
> 
> ```diff
> diff --git a/nixos/modules/system/boot/systemd/initrd.nix b/nixos/modules/system/boot/systemd/initrd.nix
> index 3f40a5b2dfa0..7476a4f9040a 100644
> --- a/nixos/modules/system/boot/systemd/initrd.nix
> +++ b/nixos/modules/system/boot/systemd/initrd.nix
> @@ -57,7 +57,6 @@ let
>      "systemd-ask-password-console.service"
>      "systemd-fsck@.service"
>      "systemd-halt.service"
> -    "systemd-hibernate-resume@.service"
>      "systemd-journald-audit.socket"
>      "systemd-journald-dev-log.socket"
>      "systemd-journald.service"
> diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
> index 19226904e0e0..6ca73349ccb4 100644
> --- a/pkgs/top-level/all-packages.nix
> +++ b/pkgs/top-level/all-packages.nix
> @@ -28515,6 +28515,7 @@ with pkgs;
>      withTpm2Tss = false;
>      withUserDb = false;
>      withUkify = false;
> +    withBootloader = false;
>    };
>    systemdStage1 = systemdMinimal.override {
>      pname = "systemd-stage-1";
> ```

Applied. @flokli Patches reflowed according to the guidelines, let me know if I messed up something.

ElvishJerricco commented:
```
$ sudo env NIXOS_INSTALL_BOOTLOADER=1 /nix/var/nix/profiles/system/bin/switch-to-configuration boot 
Copied "/nix/store/9qpbi9wp8rvy6mjlv06f6c7nxwdp1938-systemd-254/lib/systemd/boot/efi/systemd-bootx64.efi" to "/boot/EFI/systemd/systemd-bootx64.efi".
Copied "/nix/store/9qpbi9wp8rvy6mjlv06f6c7nxwdp1938-systemd-254/lib/systemd/boot/efi/systemd-bootx64.efi" to "/boot/EFI/BOOT/BOOTX64.EFI".
⚠️ Mount point '/boot' which backs the random seed file is world accessible, which is a security hole! ⚠️
⚠️ Random seed file '/boot/loader/random-seed' is world accessible, which is a security hole! ⚠️
Random seed file /boot/loader/random-seed successfully refreshed (32 bytes).
Created EFI boot entry "Linux Boot Manager".
```

Looks like we've got a new warning to account for in `systemd-boot-builder.py`

ElvishJerricco commented:
Also, I'm not sure why yet, but it appears the 254 update has caused `/dev/zfs` to have `0600` permissions instead of the usual `0666`, so e.g. regular users can't `zpool status` now.

RaitoBezarius committed:
python3Packages.systemd: add raitobezarius as a maintainer

It is pretty important package and no one seems to care about it anymore. :)

RaitoBezarius committed:
python3Packages.systemd: ignore tests that uses a direct system open call for /etc/machine-id

systemd recommends anyway to ignore them too.

RaitoBezarius commented:
@nlewo I'm seeing osqueryd issues with systemd v254:
```
août 04 00:47:07 core01 osqueryd[973941]: osqueryd started [version=5.5.1]
août 04 00:47:07 core01 osqueryd[973943]: E0804 00:47:07.948038 973943 init.cpp:700] An error occurred during extension manager startup: Extension socket directory missing: /var/osquery/osquery.em
août 04 00:47:07 core01 osqueryd[973943]: E0804 00:47:07.949370 973943 shutdown.cpp:79] Cannot activate filesystem logger plugin: Could not create file: /var/log/osquery/osqueryd.results.log
août 04 00:47:07 core01 osqueryd[973943]: Cannot activate filesystem logger plugin: Could not create file: /var/log/osquery/osqueryd.results.log
août 04 00:47:10 core01 osqueryd[973941]: E0804 00:47:10.672145 973942 shutdown.cpp:79] Worker returned exit status
août 04 00:47:10 core01 osqueryd[973941]: Worker returned exit status
août 04 00:47:10 core01 systemd[1]: osqueryd.service: Main process exited, code=exited, status=78/CONFIG
```

Please take a look if possible.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/29

RaitoBezarius commented:
Additional bug during a ZFS unlock in systemd stage 1: keymap went back to `us` instead of the specified `fr-latin1`.

ElvishJerricco commented:
So here are the things I see needing to still be done:

- We need to address the change in hibernation behavior from systemd. Now, if you don't use EFI, systemd will refuse to hibernate unless you have a `resume=` cmdline parameter. If you do use EFI, systemd will set some EFI variables, expecting them to be used for resume (unless `resume=` is present), which we don't respect. We could use these instead of the logic we use to auto-detect the hibernate image device.
- We need a solution to https://github.com/systemd/systemd/issues/28653. I know that issue is about ZFS specifically, so it seems like we could add a simple workaround, but it's actually a much broader problem, so we should fix it completely. We can either revert https://github.com/systemd/systemd/commit/b42482af904ae0b94a6e4501ec595448f0ba1c06, or we can backport their fix https://github.com/systemd/systemd/pull/28784#issuecomment-1685238230
- We have an issue with vconsole not working right with systemd stage 1. This is because the udev rules now start `systemd-vconsole-setup.service` instead of whatever they did before. Luckily all this means is that now we have to add that service to `boot.initrd.systemd.additionalUpstreamUnits` in `nixos/modules/config/console.nix`.
- Finally, we still have the warning about globally readable `/boot` to address. I'm not entirely sure what to do here. Maybe change some permissions in `systemd-boot-builder.py`?

ElvishJerricco commented:
There's also some majorly quirky nuance with respect to systemd stage 1 and the new hibernation situation. So far, we've simply not supported not setting `boot.resumeDevice` (a.k.a. `resume=`) when using hibernation and systemd stage 1. We could simply continue this practice. But if we want to support systemd's expected behavior around these new EFI variables, we need to use an EFI-enabled systemd build in stage 1, and we need to include the `efivarfs` kernel module in initrd.

This does create a minor snag though. We currently have `systemd-gpt-auto-generator` in the initrd, and once `efivarfs` is there and an EFI-enabled build is used, that generator will take over the rootfs unless you have `root=` on the cmdline or `rd.systemd.gpt_auto=false`. This is a self imposed snag though; we can just remove the generator, since we don't actually use it.

By eliminating that snag and enabling EFI in systemd stage 1, we can drop the requirement for `boot.resumeDevice` with systemd stage 1 as long as the user has EFI! Which, now that systemd mandates the same anyway, is true for scripted initrd as well, bringing them to parity.

I would like to have scripted initrd check those EFI variables and use the `resume=` parameter rather than the way it currently does everything, but we can save that for later.

RaitoBezarius commented:
> There's also some majorly quirky nuance with respect to systemd stage 1 and the new hibernation situation. So far, we've simply not supported not setting `boot.resumeDevice` (a.k.a. `resume=`) when using hibernation and systemd stage 1. We could simply continue this practice. But if we want to support systemd's expected behavior around these new EFI variables, we need to use an EFI-enabled systemd build in stage 1, and we need to include the `efivarfs` kernel module in initrd.
> 
> This does create a minor snag though. We currently have `systemd-gpt-auto-generator` in the initrd, and once `efivarfs` is there and an EFI-enabled build is used, that generator will take over the rootfs unless you have `root=` on the cmdline or `rd.systemd.gpt_auto=false`. This is a self imposed snag though; we can just remove the generator, since we don't actually use it.
> 
> By eliminating that snag and enabling EFI in systemd stage 1, we can drop the requirement for `boot.resumeDevice` with systemd stage 1 as long as the user has EFI! Which, now that systemd mandates the same anyway, is true for scripted initrd as well, bringing them to parity.
> 
> I would like to have scripted initrd check those EFI variables and use the `resume=` parameter rather than the way it currently does everything, but we can save that for later.

I'm not sure if I understand well your proposal.

Can you clarify if you propose to:

(1) Use an EFI enabled build for systemd (which we do in this PR)
(2) Add `efivarfs` in initrd (which we don't)
(3) Set `rd.systemd_gpt_auto=false` always or not?
(4) Dropping `boot.resumeDevice` need?
(5) Changing scripted initrd to adhere to the new logic you propose at the end in another PR.

Am I right?

RaitoBezarius committed:
nixos/console: use systemd-vconsole-setup.service from upstream for sd initrd

This fixes a bug where the vconsole was not working as intended in systemd stage 1 with systemd v254.

udev rules are now starting with this service instead of whatever happened before.

RaitoBezarius committed:
systemd: apply upstream patch for tmpfiles

This fixes notably the fact that /dev/zfs was not usable anymore as a user,
and potentially other things.

Tracked in systemd upstream under issue number 28653, 28765.

RaitoBezarius commented:
> So here are the things I see needing to still be done:
> 
> * We need to address the change in hibernation behavior from systemd. Now, if you don't use EFI, systemd will refuse to hibernate unless you have a `resume=` cmdline parameter. If you do use EFI, systemd will set some EFI variables, expecting them to be used for resume (unless `resume=` is present), which we don't respect. We could use these instead of the logic we use to auto-detect the hibernate image device.

I would love if you can push a commit for this one, I am afraid I don't have all the context you had on this one.

> * We need a solution to [permissions issues with openzfs systemd/systemd#28653](https://github.com/systemd/systemd/issues/28653). I know that issue is about ZFS specifically, so it seems like we could add a simple workaround, but it's actually a much broader problem, so we should fix it completely. We can either revert [systemd/systemd@b42482a](https://github.com/systemd/systemd/commit/b42482af904ae0b94a6e4501ec595448f0ba1c06), or we can backport their fix [Udev vs tmpfiles take 3 systemd/systemd#28784 (comment)]
(https://github.com/systemd/systemd/pull/28784#issuecomment-1685238230)

Implemented.

> * We have an issue with vconsole not working right with systemd stage 1. This is because the udev rules now start `systemd-vconsole-setup.service` instead of whatever they did before. Luckily all this means is that now we have to add that service to `boot.initrd.systemd.additionalUpstreamUnits` in `nixos/modules/config/console.nix`.

Implemented.

> * Finally, we still have the warning about globally readable `/boot` to address. I'm not entirely sure what to do here. Maybe change some permissions in `systemd-boot-builder.py`?

I guess we can fixup the permissions, but that won't work for, e.g. lanzaboote. Maybe this should be a user activation script for fixup, but heh.

ElvishJerricco commented:
> > * We need a solution to [permissions issues with openzfs systemd/systemd#28653](https://github.com/systemd/systemd/issues/28653). I know that issue is about ZFS specifically, so it seems like we could add a simple workaround, but it's actually a much broader problem, so we should fix it completely. We can either revert [systemd/systemd@b42482a](https://github.com/systemd/systemd/commit/b42482af904ae0b94a6e4501ec595448f0ba1c06), or we can backport their fix [Udev vs tmpfiles take 3 [Udev vs tmpfiles take 3 systemd/systemd#28784](https://github.com/systemd/systemd/pull/28784) (comment)]
> >   ([Udev vs tmpfiles take 3 systemd/systemd#28784 (comment)](https://github.com/systemd/systemd/pull/28784#issuecomment-1685238230))
> 
> Implemented.

I believe backporting their fix requires adding a new unit `systemd-tmpfiles-setup-dev-early.service` to the default upstream units.

> I'm not sure if I understand well your proposal.
> 
> Can you clarify if you propose to:
> 
> (1) Use an EFI enabled build for systemd (which we do in this PR)
> (2) Add `efivarfs` in initrd (which we don't)
> (3) Set `rd.systemd_gpt_auto=false` always or not?
> (4) Dropping `boot.resumeDevice` need?
> (5) Changing scripted initrd to adhere to the new logic you propose at the end in another PR.
> 
> Am I right?

1. We do not currently use an EFI enabled build for systemd *in stage 1*.
2. Right
3. No, I think it would work to get rid of `systemd-gpt-auto-generator` instead, because we don't use that and actually actively don't want it doing anything anyway.
4. Sort of. We would be *adding* that as a requirement for non-EFI systems, because that's now required by systemd. But with the previous points, we can *remove it* as a requirement _for systemd stage 1_ on EFI systems.
5. Yea, I think that would be another PR.

ElvishJerricco commented:
Alternatively we can just take no action now and make improvements later. No matter what we do, we have to tighten the constraint and say that non-EFI users must have `boot.resumeDevice` set for hibernation to work. Otherwise, everything else is status quo.

nikstur commented:
> because we don't use that and actually actively don't want it doing anything anyway.

I do use it and we should at least add an option to keep it

RaitoBezarius commented:
Can someone test the NixOS containers? I do not use this feature and do not really care much about it, but it would be nice to verify we are not breaking it.

arianvp commented:
@GrahamcOfBorg test containers

arianvp commented:
@GrahamcOfBorg test containers-bridge containers-custom-pkgs containers-ephemeral containers-extra_veth containers-hosts containers-imperative containers-ip containers-macvlans containers-names containers-nested containers-physical_interfaces containers-portforward containers-reloadable containers-restart_networking containers-tmpfs containers-unified-hierarchy

arianvp commented:
Oh of course ofborg won't work as this targets staging.. sigh. not sure how to cancel these jobs now

RaitoBezarius commented:
@cole-h Did we do a big mistake? Let us know if we can cancel them, apologies for the inconveniences.

cole-h commented:
I wouldn't worry about it :)

(Mostly because there's no way to cancel them :)

ElvishJerricco commented:
@nikstur How are you using it? Without patching nixpkgs, NixOS currently generates errors if you try to configure gpt-auto for the root fs. If you're not talking about the root fs, then we're talking about two different things.

nikstur commented:
> If you're not talking about the root fs, then we're talking about two different things.

Yes, you're right. Although I'd like to potentially do that (although as you said we'd need to remove the error when you don't set `fileSystems."/"` first.)

Nyabinary commented:
Whats the status of this

RaitoBezarius commented:
> Whats the status of this

As you can see, some points are left to be solved, anyone is free to PR them, untangle them. We coordinate in the NixOS systemd Matrix room.

As you can also see on my profile, I'm on vacations for now :).

nikstur committed:
systemd: 254 -> 254.3

nikstur commented:
I fixed this PR up to use 254.3 and fixed the conflicts here: https://github.com/nikstur/nixpkgs/tree/systemd-254

As far as I understand the comments in this history, we should be good to go.

@RaitoBezarius please let me know how you want me to incorporate these changes into your branch.

We have these action items on our todo list:

- Improve the hibernation/resume situation
- Improve the permission situation on `/boot`

alyssais commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -192,7 +199,6 @@ stdenv.mkDerivation (finalAttrs: {
     ./0017-core-don-t-taint-on-unmerged-usr.patch
     ./0018-tpm2_context_init-fix-driver-name-checking.patch
-    ./0019-bootctl-also-print-efi-files-not-owned-by-systemd-in.patch
   ] ++ lib.optional stdenv.hostPlatform.isMusl (
```
[Updated musl patches are available.](https://git.openembedded.org/openembedded-core/tree/meta/recipes-core/systemd/systemd_254.bb?id=69d37620a7a0626eaee1c0b37057be2ac2687290#n35)  Up to you whether you want to include them here or leave it for somebody else to fix during staging as follow up.

nikstur replied:
Never touched the musl stuff, so I'd leave it to someone else.

RaitoBezarius replied:
I will pull them.

RaitoBezarius replied:
I tried: `pkgsMusl.systemdMinimal` compiles, `pkgsMusl.systemd` does not, here's the diff:

```diff
diff --git a/pkgs/os-specific/linux/systemd/default.nix b/pkgs/os-specific/linux/systemd/default.nix
index e3e95246a0bf..d14ac5880f63 100644
--- a/pkgs/os-specific/linux/systemd/default.nix
+++ b/pkgs/os-specific/linux/systemd/default.nix
@@ -204,16 +204,16 @@ stdenv.mkDerivation (finalAttrs: {
   ] ++ lib.optional stdenv.hostPlatform.isMusl (
     let
       oe-core = fetchzip {
-        url = "https://git.openembedded.org/openembedded-core/snapshot/openembedded-core-f34f6ab04b443608497b73668365819343d0c2fe.tar.gz";
-        sha256 = "DFcLPvjQIxGEDADpP232ZRd7cOEKt6B48Ah29nIGTt4=";
+        url = "https://git.openembedded.org/openembedded-core/snapshot/openembedded-core-69d37620a7a0626eaee1c0b37057be2ac2687290.tar.gz";
+        hash = "sha256-oq7hTDdcimBxyq3qPAbqv5GP8rkygkyda63SFNXxXmk=";
       };
       musl-patches = oe-core + "/meta/recipes-core/systemd/systemd";
     in
     [
       (musl-patches + "/0001-Adjust-for-musl-headers.patch")
+      (musl-patches + "/0003-errno-util-Make-STRERROR-portable-for-musl.patch")
       (musl-patches + "/0005-pass-correct-parameters-to-getdents64.patch")
       (musl-patches + "/0006-test-bus-error-strerror-is-assumed-to-be-GNU-specifi.patch")
-      (musl-patches + "/0007-Add-sys-stat.h-for-S_IFDIR.patch")
       (musl-patches + "/0009-missing_type.h-add-comparison_fn_t.patch")
       (musl-patches + "/0010-add-fallback-parse_printf_format-implementation.patch")
       (musl-patches + "/0011-src-basic-missing.h-check-for-missing-strndupa.patch")
@@ -229,7 +229,11 @@ stdenv.mkDerivation (finalAttrs: {
       (musl-patches + "/0022-Handle-__cpu_mask-usage.patch")
       (musl-patches + "/0023-Handle-missing-gshadow.patch")
       (musl-patches + "/0024-missing_syscall.h-Define-MIPS-ABI-defines-for-musl.patch")
-      (musl-patches + "/0026-src-boot-efi-efi-string.c-define-wchar_t-from-__WCHA.patch")
+      (musl-patches + "/0025-include-sys-file.h-for-LOCK_EX.patch")
+      (musl-patches + "/0026-test-test-sizeof-Include-sys-timex.h-for-struct-time.patch")
+      (musl-patches + "/0027-include-missing-sys-file.h-for-LOCK_EX.patch")
+      (musl-patches + "/0028-sd-event-Make-malloc_trim-conditional-on-glibc.patch")
+      (musl-patches + "/0029-shared-Do-not-use-malloc_info-on-musl.patch")
     ]
   );
```

RaitoBezarius replied:
If you can pick it up from there in a follow-up, that'd be awesome!

RaitoBezarius replied:
It seems BTW we are lacking:

> -      (musl-patches + "/0026-src-boot-efi-efi-string.c-define-wchar_t-from-__WCHA.patch")

as the error is related to wchar redefinition stuff.

RaitoBezarius committed:
systemd: add release notes for v254

We add information for 23.11 regarding the v254 release.
This information may be modified before the actual release.

nikstur committed:
nixos/image: fix for systemd 254

nikstur committed:
systemdStage1: re-include repart

alyssais commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -371,7 +368,7 @@ stdenv.mkDerivation (finalAttrs: {
   + ''
     shopt -s extglob
-    patchShebangs tools test src/!(rpm|kernel-install|ukify) src/kernel-install/test-kernel-install.sh
+    patchShebangs tools test src/!(rpm|ukify) src/kernel-install/test-kernel-install.sh
```
What was the reason for this change?  It broke cross compilation (because these scripts end up with a build bash shebang), and isn't mentioned in the commit message.

nikstur replied:
My bad for not making this explicit. But for all I know, `kernel-install` is not a script anymore. It was rewritten in C. What am I missing?

alyssais replied:
[All the .install shell scripts.](https://github.com/systemd/systemd/tree/main/src/kernel-install)

RaitoBezarius replied:
I am seeing
```
error: build of '/nix/store/llnwcn2bgri0m9cn58h8ym6wj42s1jyv-systemd-minimal-aarch64-unknown-linux-gnu-254.3.drv' on 'ssh://build01-x86-newtype' failed: output '/nix/store/7p5kh7smhwni9pwihs3r4mnzm5sxr022-systemd-minimal-aarch64-unknown-linux-gnu-254.3' is not allowed to refer to the following paths:
         /nix/store/vqvj60h076bhqj6977caz0pfxs6543nb-bash-5.2-p15
```

nikstur replied:
Could it be related to this: https://github.com/NixOS/nixpkgs/pull/252874 ?

nikstur replied:
Applying these patches seems to fix it.

RaitoBezarius replied:
So we need to remove the bash path in some out thingie

lilyinstarlight replied:
No, this line just needs to be put back as it was before since the .install scripts get installed, as Alyssa mentioned, and need to use the hosr bash, not the build bash

nikstur replied:
I'll open a PR

nikstur replied:
https://github.com/NixOS/nixpkgs/pull/255165

RaitoBezarius replied:
Thank you @alyssais for the ping and sorry for the random drop

RaitoBezarius replied:
(and thank you @nikstur for fast turnaround in an ICE)

misuzu commented:
systemd on staging-next is broken on armv7l-linux https://hydra.armv7l.xyz/build/17881/nixlog/2
```
[662/1780] Generating src/boot/efi/addonarm.efi.stub with a custom command
FAILED: src/boot/efi/addonarm.efi.stub 
/build/source/tools/elf2efi.py --version-major=254 --version-minor=0 --efi-major=1 --efi-minor=1 --subsystem=10 --minimum-sections=15 src/boot/efi/addonarm.elf.stub src/boot/efi/addonarm.efi.stub
Traceback (most recent call last):
  File "/build/source/tools/elf2efi.py", line 601, in <module>
    main()
  File "/build/source/tools/elf2efi.py", line 597, in main
    elf2efi(parser.parse_args())
  File "/build/source/tools/elf2efi.py", line 499, in elf2efi
    pe_reloc_s = convert_elf_relocations(elf, opt, sections)
  File "/build/source/tools/elf2efi.py", line 388, in convert_elf_relocations
    raise RuntimeError("ELF .dynamic section is missing.")
RuntimeError: ELF .dynamic section is missing.
[663/1780] Linking target src/boot/efi/linuxarm.elf.stub
[664/1780] Compiling C object src/import/libimport-common.a.p/import-common.c.o
[665/1780] Compiling C object src/boot/efi/systemd-bootarm.elf.p/boot.c.o
ninja: build stopped: subcommand failed.
```

alyssais commented:
Cross to riscv64 is broken for the same reason:

```
FAILED: src/boot/efi/addonriscv64.efi.stub
/build/source/tools/elf2efi.py --version-major=254 --version-minor=0 --efi-major=1 --efi-minor=1 --subsystem=10 --minimum-sections=15 src/boot/efi/addonriscv64.elf.stub src/boot/efi/addonriscv64.efi.stub
Traceback (most recent call last):
  File "/build/source/tools/elf2efi.py", line 601, in <module>
    main()
  File "/build/source/tools/elf2efi.py", line 597, in main
    elf2efi(parser.parse_args())
  File "/build/source/tools/elf2efi.py", line 499, in elf2efi
    pe_reloc_s = convert_elf_relocations(elf, opt, sections)
  File "/build/source/tools/elf2efi.py", line 388, in convert_elf_relocations
    raise RuntimeError("ELF .dynamic section is missing.")
RuntimeError: ELF .dynamic section is missing.
```

nikstur commented:
Is this an upstream error or something that we have messed up?

misuzu commented:
> Is this an upstream error or something that we have messed up?

I think it's an upstream issue. Likely related: https://github.com/systemd/systemd/issues/27821, https://github.com/systemd/systemd/pull/26641

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/33

lopsided98 commented:
Arch Linux ARM builds `addonarm.efi.stub` successfully so I don't think this is an upstream issue.

lopsided98 commented:
The root cause of this bug appears to be that gcc doesn't support the `-static-pie` argument on 32-bit ARM or RISC-V. The argument is accepted, but doesn't actually do anything. On x86_64, [`-static-pie` causes `-static -pie --no-dynamic-linker -z text`](https://github.com/gcc-mirror/gcc/blob/fd35d72a3dcd5ba14d81a1890236acd0145497e1/gcc/config/i386/gnu-user64.h#L67) to be passed to the linker (with slight variations on the other supported architectures).

In systemd, they noticed something wasn't quite right with `-static-pie`, and therefore they manually pass most of these arguments to the linker, see:
https://github.com/systemd/systemd/blob/cb88da825460c410139a9a3041af849ee2d01d5b/src/boot/efi/meson.build#L156-L158

Notably though, they don't pass `-pie`. The reason for this is that pretty much every distribution except for nixpkgs builds gcc with `--enable-default-pie`, so `-pie` is passed to the linker even without `-static-pie`.

This raises several questions about how this should be fixed:
* If manually passing `-pie --no-dynamic-linker -z text` is enough to make static PIE work, why doesn't gcc support `-static-pie` on more architectures?
* Should systemd pass `-Wl,-pie` to work around this gcc issue?
* Why don't we build gcc with `--enable-default-pie` in nixpkgs? We have the `pie` hardening option that is disabled by default (except on musl) because it apparently caused problems. Would these same problems occur with `--enable-default-pie`?

lopsided98 commented:
It looks like `-static-pie` support on the gcc side is as simple as passing those flags to the linker. Some more work is probably needed in glibc to support it, but that is not relevant to systemd.

In fact, someone submitted a patch to gcc to add `-static-pie` support on ARM, but it was ignored: https://patchwork.ozlabs.org/project/gcc/patch/c7dc02a4-5fc1-956b-6d89-6c09a1a48b5a@gmail.com/

saschagrunert commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -146,17 +149,21 @@ assert withCoredump -> withCompression;
 assert withHomed -> withPam;
 assert withUkify -> withEfi;
+assert withRepart -> withCryptsetup;
+assert withBootloader -> withEfi;
```
This breaks cross compilation for powerpc:

```
> nix-build --arg crossSystem '{ config = "powerpc64le-unknown-linux-gnu"; }' -A systemd
error:
       … while evaluating a branch condition

         at /home/sascha/git/nixpkgs/lib/customisation.nix:86:7:

           85|     in
           86|       if builtins.isAttrs result then
             |       ^
           87|         result // {

       … while calling the 'isAttrs' builtin

         at /home/sascha/git/nixpkgs/lib/customisation.nix:86:10:

           85|     in
           86|       if builtins.isAttrs result then
             |          ^
           87|         result // {

       error: assertion '(withBootloader -> withEfi)' failed

       at /home/sascha/git/nixpkgs/pkgs/os-specific/linux/systemd/default.nix:153:1:

          152| assert withRepart -> withCryptsetup;
          153| assert withBootloader -> withEfi;
             | ^
          154| # passwdqc is not packaged in nixpkgs yet, if you want to fix this, please submit a PR.
```

RaitoBezarius replied:
Is `stdenv.hostPlatform.isEfi` true for PowerPC? If not, I think we should disable the bootloader for PowerPC.

lopsided98 replied:
Same problem on armv5tel

lopsided98 replied:
These are all the platforms where `isEfi` is true:

https://github.com/NixOS/nixpkgs/blob/0c19eb6916ade1024700a42ef72ba69656053f3c/lib/systems/inspect.nix#L94-L100

It seems like `withBootloader` should default to the value of `withEfi`.

lopsided98 replied:
https://github.com/NixOS/nixpkgs/pull/259510
